Implement Keycloak authentication

lunasy26 opened the pull request.

Commit:
feat: implement Keycloak authentication and authorization

- Integrated Keycloak for user authentication and authorization
- Configured AddOpenIdConnect method for seamless Keycloak interaction
- Enhanced UI with login/logout options and visual feedback
- Enabled external access to Keycloak from Docker container using host.docker.internal

Commit:
feat: change README

Commit:
fix readme

Commit:
second fix README

Kharlap-Sergey commented on src/Presentation/Memo4me.UI/Program.cs:
```diff
@@ -1,35 +1,35 @@
-public class Program
+var builder = WebApplication.CreateBuilder(args); 
+{
+    var configuration = builder.Configuration;
```
why can't we pass it directly?

Kharlap-Sergey commented on src/Presentation/Memo4me.UI/Program.cs:
```diff
@@ -1,35 +1,35 @@
-
-public class Program
+var builder = WebApplication.CreateBuilder(args); 
+{
```
Why do we need new code block here

lunasy26 replied:
for visual convenience
If this is wrong, I'll fix it

Kharlap-Sergey replied:
It is ok, I just never seen this so it baffles the eye

Kharlap-Sergey commented on src/Presentation/Memo4me.UI/Components/Pages/Logout.razor:
```diff
@@ -0,0 +1,23 @@
+        var idToken = await httpContext.GetTokenAsync("id_token");
+        await httpContext.SignOutAsync();
+        
+        var keycloakLogoutUrl = $"http://host.docker.internal:7270/realms/MemoRealm/protocol/openid-connect/logout" +
```
hardcoded url

lunasy26 replied:
fixed

Kharlap-Sergey commented on src/Presentation/Memo4me.UI/Components/Pages/Login.razor:
```diff
@@ -0,0 +1,19 @@
+<h3>Login</h3>
+
+@code {
+    protected override async Task OnInitializedAsync()
```
Could you please explain to me, when and in which conditions this method is invoked

lunasy26 replied:
OnInitializedAsync() is a lifecycle method of the blazor component that is called once when the component is initialized and after it has been added to rendering
In my case i use this method to initialize the authentication process, sending the user to the login page

Kharlap-Sergey commented on src/Presentation/Memo4me.UI/Extensions/AddAuthOpenIdExtension.cs:
```diff
@@ -0,0 +1,32 @@
+                options.ClientSecret = configuration["Authentication:OpenIdConnect:ClientSecret"];
+                options.ResponseType = configuration["Authentication:OpenIdConnect:ResponseType"];
+                options.SaveTokens = true;
+                options.RequireHttpsMetadata = false;
```
should be configurable, and set to true by default, as it critical for production

lunasy26 replied:
done

Kharlap-Sergey commented on src/Presentation/Memo4me.UI/appsettings.json:
General recommendation
appsettings.json should contain schema with values required for production, if you need something for development use appsettings.Dev...json 
Secret information shouldn't be stored here at all
When run in VS there is UserSecrets configuration provider, but when run through docker you don't have such an option so it better to provider them through env variables

Commit:
refactor: externalize authentication settings and improve security

- Moved sensitive OpenIdConnect settings (e.g., ClientSecret) to environment variables for better security.
- Configured `RequireHttpsMetadata` to automatically load from appsettings based on environment (Development or Production).
- Simplified Keycloak logout URL generation to support dynamic configuration.
